WordPress/wordpress-playground
# Blueprints sidebar section for single-click Playground presets

## Motivation for the change, related issues

Adds the Blueprint Gallery to the Playground UI as a first-class feature.

![CleanShot 2024-10-10 at 13 40 17@2x](https://github.com/user-attachments/assets/4e30e0e6-da9b-4dfa-9b21-6d15b7c0e44b)


## Implementation

### Sourcing the list of Blueprints

GitHub automation maintains a JSON index of all the Blueprints in the directory at https://raw.githubusercontent.com/WordPress/blueprints/trunk/index.json. This PR just sources that data and adds a DataView where it can be browsed.

It ships a new `useFetch()` React hook that makes the state management around `fetch()` easier. Maybe this simplistic  hook will be enough, or maybe we'll quickly replace it with `react-query` or so. We'll see.

### Fetch proxy for offline caching

This PR ships a "proxy" service worker route for offline caching of cross-origin requests.

Playground is now an offline-enabled PWA and anytime we rely on network resources, we must consider what happens when there's network connectivity.

For same-origin request, that's easy. We can just cache them in the `fetch` event handler in the service worker. However, that event handler won't handle cross-origin requests. Therefore, this PR ships a special same-origin `/proxy/network-first-fetch/` route that can cache cross-origin requests in `CacheStorage`.

For example, the following request would fetch the list of all the Blueprints from the Blueprints directory:

```
https://playground.wordpress.net/proxy/network-first-fetch/https://raw.githubusercontent.com/WordPress/blueprints/trunk/index.json
```

The service worker would be notified, then it would strip the `https://playground.wordpress.net/proxy/network-first-fetch/` prefix, and source the actual information from `https://raw.githubusercontent.com/WordPress/blueprints/trunk/index.json`. If the network is available, it will always request the fresh data. If the network is not available, it would source the data from `CacheStorage` if available.

## Testing Instructions

Interact with the gallery, confirm it works in the offline mode, confirm the blueprints load without issues and that the rest of the UI works the same as before this PR.

## Follow-up work 

- [ ] Revisit the design based on user feedback and @jarekmorawski's explorations in [Figma](https://www.figma.com/design/VmNXsvW4mKyKQecQUIUZm4/%5BProject%5D-Playground-2.0?node-id=0-1)
- [ ] Integrate with the Blueprints builder when it's ready
- [ ] Add GitHub connection to make contributing Blueprints easy
- [ ] Integrate with the upcoming visual Blueprints builder

cc @brandonpayton @bgrgicak @mtias
